docs/benefits-over-pyright/fixed-context-manager-exit-types.md

```diff
@@ -0,0 +1,140 @@
+        traceback: TracebackType | None,
+        /,
+    ) -> Literal[False]:
+        print("exiting context manager")
+        return False
+```
+
+pyright will take this into account when determining reachability:
+
+```py
+def raise_exception() -> Never:
+    raise Exception
+
+with SuppressError():
+    foo: int = raise_exception()
+
+# when the exception is raised, the context manager exits before foo is assigned to:
+print(foo)  # error: "foo" is unbound (reportPossiblyUnboundVariable)
+```
+
+```py
+with Log():
+    foo: int = raise_exception()
+
+# when the exception is raised, it does not get suppressed so this line can never run:
+print(foo)  # error: Code is unreachable (reportUnreachable)
+```
+
+however, due to [a bug in mypy](https://github.com/python/mypy/issues/8766) that [pyright blindly copied and accepted as the "standard"](https://github.com/microsoft/pyright/issues/6034#issuecomment-1738941412), a context manager will incorrectly be treated as if it never suppresses exceptions if its return type is a union of `bool | None`:
+
+```py
+class SuppressError(AbstractContextManager[None, bool | None]):
+    @override
+    def __enter__(self) -> None:
+        pass
+
+    @override
+    def __exit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_value: BaseException | None,
+        traceback: TracebackType | None,
+        /,
+    ) -> bool | None:
+        return True
+
+
+with SuppressError():
+    foo: int = raise_exception()
+
+# this error is wrong because this line is actually reached at runtime:
+print(foo)  # error: Code is unreachable (reportUnreachable)
+```
+
+## the solution
+
+basedpyright introduces a new setting, `strictContextManagerExitTypes` to address this issue. when enabled, context managers where the `__exit__` dunder returns `bool | None` are treated the same way as context managers that return `bool` or `Literal[True]`. put simply, if `True` is assignable to the return type, then it's treated as if it can suppress exceptions.
+
+## issues with `@contextmanager`
+
+the reason we support disabling this fix using the `strictContextManagerExitTypes` setting is because it will cause all context managers decorated with `@contextlib.contextmanager` to be treated as if they can suppress an exception, even if they never do:
+
+```py
+@contextmanager
+def log():
+    print("entering context manager")
+    try:
+        yield
+    finally:
+        print("exiting context manager")
+
+with log():
+    foo: int = get_value()
+
+# basedpyright accounts for the possibility that get_value raised an exception and foo
+# was never assigned to, even though this context manager never suppresses exceptions
+print(foo)  # error: "foo" is unbound (reportPossiblyUnboundVariable)
+```
+
+this is because there's no way to tell a type checker whether the function body contains a `try`/`except` statement, which is necessary to suppress exeptions when using the `@contextmanager` decorator:
```

Comment:
technically, the context manager could `yield from` a different generator, but that's a very much an edge case:
```py
from contextlib import contextmanager


def f():
    try:
        yield
    finally:
        return True

@contextmanager
def c():
    yield from f()

with c():
    raise Exception("oops")

print("done")
```

Reply:
i think thats implied